rust_src/src/fonts.rs

```diff
@@ -265,4 +310,51 @@ pub fn font_at_lisp(
     unsafe { font_at(-1, pos, ptr::null_mut(), w.as_mut(), string) }
 }
 
+/// List available fonts matching FONT-SPEC on the current frame.
+/// Optional 2nd argument FRAME specifies the target frame.
+/// Optional 3rd argument NUM, if non-nil, limits the number of returned fonts.
+/// Optional 4th argument PREFER, if non-nil, is a font-spec to
+/// control the order of the returned list.  Fonts are sorted by
+/// how close they are to PREFER.
+#[lisp_fn(min = "1")]
+pub fn list_fonts(
+    font_spec: LispFontSpecRef,
+    frame: LispFrameLiveOrSelected,
+    num: Option<LispNumber>,
+    prefer: Option<LispFontSpecRef>,
+) -> LispObject {
+    let mut frame: LispFrameRef = frame.into();
+
+    let n = match num.map(|n| n.to_fixnum()) {
+        Some(n) if n < 0 => return Qnil,
+        Some(n) => n,
+        None => 0,
+    } as usize;
+
+    let list = unsafe { font_list_entities(frame.as_mut(), font_spec.into()) };
+    match list.as_cons() {
+        Some(list) => {
+            if list.cdr().is_nil() && list.car().force_vector().len() == 1 {
```

Comment:
`force_vector` gets called twice here. Is there a way of pattern matching it?

Reply:
It should be possible with an `if let`

Reply:
Turns out you can do it even more neatly with a guarded match.